### PR TITLE
Some updates from SPL work 

### PR DESCRIPTION
**Pull request by Mehrad0711**
Most changes are around better handling of string parameters for augmentation.
Also, updating requote and augmentation scripts to accept examples with multiple TT annotations.

### CODE REVIEW

File: tool/autoqa/schemaorg/manual-annotations.js [hunk 1]
```diff
@@ -563,12 +563,12 @@ const STRUCT_INCLUDE_THING_PROPERTIES = new Set([
...
 
 const STRING_FILE_OVERRIDES = {
-    'org.schema.Restaurant:Restaurant_name': 'org.openstreetmap:restaurant',
+    // 'org.schema.Restaurant:Restaurant_name': 'org.openstreetmap:restaurant',
     'org.schema.Person:Person_name': 'tt:person_full_name',
     'org.schema.Person:Person_alumniOf': 'tt:university_names',
     'org.schema.Person:Person_worksFor': 'tt:company_name',
     'org.schema.Person:Person_jobTitle': 'tt:job_title',
-    'org.schema.Hotel:Hotel_name': 'org.openstreetmap:hotel',
+    // 'org.schema.Hotel:Hotel_name': 'org.openstreetmap:hotel',
```
Comment (gcampax) [inline tool/autoqa/schemaorg/manual-annotations.js]: What's special about Restaurant and Hotel? What about all the other domains? Your scripts should remove `#[string_values]` or replace them with a localized file instead.

File: tool/requote.js [hunk 2]
```diff
@@ -51,9 +51,10 @@ function findSpanContaining(index, spansBySentencePos) {
     return false;
 }
 
+// check normalized version of tokens to handle accent marks/ grapheme properly
 function substringIsFound(sequence, substring, i) {
     for (let j = 0; j < substring.length; j++) {
-        if (sequence[i+j] !== substring[j])
+        if (sequence[i+j].normalize() !== substring[j].normalize())
```
Comment (gcampax) [inline tool/requote.js:57]: This code is run on tokenized examples (sentence/programs) so everything is NFKD already. Why is this needed?

File: lib/dataset-tools/augmentation/replace_parameters.js [hunk 3]
```diff
@@ -106,12 +106,13 @@ class UniformValueList {
...
 
 class ValueListLoader {
-    constructor(provider, samplingType, rng) {
+    constructor(provider, samplingType, subsetParamSet, rng) {
         this._provider = provider;
 
         this._emptyList = new UniformValueList([]);
         this._cache = new Map;
         this._samplingType = samplingType;
+        this._subsetParamSet = subsetParamSet;
```
Comment (gcampax) [inline lib/dataset-tools/augmentation/replace_parameters.js:115]: This option should not be a string, it should be a tuple of numbers. Do the parsing outside.

File: lib/dataset-tools/augmentation/replace_parameters.js [hunk 4]
```diff
@@ -425,6 +425,11 @@ module.exports = class ParameterReplacer {
                 continue;
             }
 
+            if (sampled === 'unspecified'){
+                attempts -= 1;
+                continue;
+            }
```
Comment (gcampax) [inline lib/dataset-tools/augmentation/replace_parameters.js:439]: What's this about? No.

File: lib/dataset-tools/augmentation/replace_parameters.js [hunk 5]
```diff
@@ -425,7 +425,7 @@ module.exports = class ParameterReplacer {
             }
 
             // fail immediately if sampled value includes return char
-            if (sampled.includes('\n')){
+            if (sampled.includes('\n') || sampled.includes('/') || sampled.includes('\\') ){
```
Comment (gcampax) [inline lib/dataset-tools/augmentation/replace_parameters.js:431]: What's going on here? Sampled is a tokenized value so it cannot have `\n`. And why is `/` or `\\` problematic? We certainly want to understand words with slashes.

File: tool/evaluate-server.js [hunk 6]
```diff
@@ -175,6 +175,12 @@ module.exports = {
             default: 0,
             help: 'Start evaluation from this line of input data',
         });
+        parser.add_argument('--output-errors', {
+            required: false,
+            default: process.stdout,
+            type: fs.createWriteStream,
+            help: "Write erred examples to this file instead of stdout"
```
Comment (gcampax) [inline tool/evaluate-server.js:182]: Btw "erred" is not a word, the word is "erroneous" (or "bogus" which is more common)

**Review by gcampax — CHANGES_REQUESTED** [review]
A lot of comments throughout but overall it's ok. I really don't like the option to skip bogus examples for evaluation though. It's way too easy to confuse "errors" with parsing errors, and there is no reason to have invalid examples in the evaluation set.

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 5 above)
Comment (Mehrad0711) [reply lib/dataset-tools/augmentation/replace_parameters.js:431]: For some languages, the tokenization was wrong for parameters with slashes in them. I will remove this and revisit later if the problem came up again.

Comment (gcampax) [reply lib/dataset-tools/augmentation/replace_parameters.js:431]: We don't split on slashes, but we do split on backslashes. At least in English. I think other languages do the same because they don't have custom rules.

File: tool/augment.js [hunk 7]
```diff
@@ -155,6 +155,12 @@ module.exports = {
             help: 'Random/ Uniform strategy assigns random/ uniform weights to parameters instead of reading from file',
             default: false
         });
+        parser.add_argument('--subset-param-set', {
+            required: false,
+            type: Array,
```
Comment (gcampax) [inline tool/augment.js:160]: This type notation is incorrect. `type` is a function that takes each value and returns the parse. You're passing one string value, and converting to array, you will get one element. Did you test this?

File: lib/dataset-tools/augmentation/replace_parameters.js [hunk 8]
```diff
@@ -330,7 +330,10 @@ module.exports = class ParameterReplacer {
     _transformValue(sentenceValue, programValue, arg) {
         if (this._requotable) {
             if (arg && arg.name === 'id' && arg.type.isEntity && arg.type.type.endsWith(':Hotel')) {
-                sentenceValue = sentenceValue.slice(0, sentenceValue.indexOf(',')).trim();
+                if (sentenceValue.startsWith('the '))
+                    sentenceValue = sentenceValue.slice('the '.length).trim();
```
Comment (gcampax) [inline lib/dataset-tools/augmentation/replace_parameters.js:334]: You should use the "definite article" stuff from langPack instead of this.

File: lib/dataset-tools/augmentation/replace_parameters.js [hunk 9]
```diff
@@ -330,7 +330,10 @@ module.exports = class ParameterReplacer {
...
+                if (sentenceValue.indexOf(',') !== -1)
+                    sentenceValue = sentenceValue.slice(0, sentenceValue.indexOf(',')).trim();
```
Comment (gcampax) [inline lib/dataset-tools/augmentation/replace_parameters.js:336]: This does not seem correct in general. Also, `,` is not a good word so it will never show up.

**Review by gcampax — CHANGES_REQUESTED** [review]
A few more comments.

File: tool/augment.js (hunk 7 above)
Comment (Mehrad0711) [reply tool/augment.js]: Ok, I changed it so we that we pass a string to augment-worker, and then parse it to an array before passing to the dataset augmenter.

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 8 above)
Comment (Mehrad0711) [reply lib/dataset-tools/augmentation/replace_parameters.js:334]: The "the" bug was coming from English side caused; where requoting was capturing "casablanca" instead of "the casablanca" for example and thus you ended up with excessive "the" everywhere. The bug is fixed and this is for cleanups on datasets generated with that bug.

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 9 above)
Comment (Mehrad0711) [reply lib/dataset-tools/augmentation/replace_parameters.js:336]: American English GoodWord accepts commas as well so that we can have locations  like "washington , dc" in params.
Some values from TripAdvisor contained repetitive/ excessive fields like " city, city, country" or "area, city, town, country". This is to clean those up.

File: tool/augment.js (hunk 7 above)
Comment (gcampax) [reply tool/augment.js]: That's fine

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 8 above)
Comment (gcampax) [reply lib/dataset-tools/augmentation/replace_parameters.js:334]: I don't quite understand.

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 9 above)
Comment (gcampax) [reply lib/dataset-tools/augmentation/replace_parameters.js:336]: It's not a good heuristic to do for everything then. Put it in make-string-datasets

File: lib/dataset-tools/augmentation/replace_parameters.js (hunk 8 above)
Comment (Mehrad0711) [reply lib/dataset-tools/augmentation/replace_parameters.js:334]: his is not for removing lang specific articles from the beginning of parameters. Same token ("the") is repeated regardless of language. I will just do this processing outside of genie and remove this here. That would create less confusion I guess.

Comment (gcampax) [reply lib/dataset-tools/augmentation/replace_parameters.js:334]: Yes please.

**Review by gcampax — APPROVED** [review]
